### PR TITLE
SWATCH-1220: Script to perform live reloads in the ephemeral environments

### PR DESCRIPTION
**Pull request by awood**
Jira issue: SWATCH-1220

## Description

This PR creates a python tool, "swatchdog", capable of performing live reloads **of the Spring** application in the ephemeral environments.

### Setup

1. Install `poetry`: `sudo dnf install poetry`
2. Create and activate the poetry environment by running `poetry shell` in the top level of the repo.
3. Install the necessary dependencies with `poetry install`
4. Log in to the ephemeral openshift cluster.
5. Build the application containers.  You will need to do a rebuild because this feature required some changes in the underlying containers.  You can use the `bin/build-images.sh` script which will try to push the images to your personal `quay.io` space and tag them with the short hash of the current HEAD.
6. Make sure the images that you pushed at the step 5 are public in Quay.io.
7. Run an EE deployment of those containers.  Here is the command I use for reference:

```shell
VER=$(git rev-parse --short=7 HEAD); QUAY_USER=awood; bonfire deploy rhsm --optional-deps-method=none \
 --no-remove-resources=all \
 -i quay.io/$QUAY_USER/swatch-system-conduit=$VER \
 -i quay.io/$QUAY_USER/swatch-producer-aws=$VER \
 -i quay.io/$QUAY_USER/rhsm-subscriptions=$VER \
 -i quay.io/$QUAY_USER/swatch-contracts=$VER \
 -p swatch-producer-red-hat-marketplace/IMAGE=quay.io/$QUAY_USER/rhsm-subscriptions \
 -p swatch-metrics/IMAGE=quay.io/$QUAY_USER/rhsm-subscriptions \
 -p swatch-subscription-sync/IMAGE=quay.io/$QUAY_USER/rhsm-subscriptions \
 -p swatch-system-conduit/IMAGE=quay.io/$QUAY_USER/rhsm-subscriptions \
 -p swatch-system-conduit/CONDUIT_IMAGE=quay.io/$QUAY_USER/swatch-system-conduit \
 -p swatch-tally/IMAGE=quay.io/$QUAY_USER/rhsm-subscriptions \
 -p swatch-producer-aws/IMAGE=quay.io/$QUAY_USER/swatch-producer-aws \
 -p swatch-api/IMAGE=quay.io/$QUAY_USER/rhsm-subscriptions \
 -p swatch-contracts/IMAGE=quay.io/$QUAY_USER/swatch-contracts \
 -p swatch-api/JAVA_OPTS_APPEND=-Dspring.devtools.restart.enabled=true \
 -p swatch-contracts/QUARKUS_LAUNCH_DEVMODE=true \
 -p swatch-contracts/JAVA_DEBUG=true
```

### Steps
1.  Make some trivial change to the Spring application.  The easiest thing to do is just add a `System.out.println("Hello world!")` in `BootApplication`.
    ```patch
    diff --git a/src/main/java/org/candlepin/subscriptions/BootApplication.java b/src/main/java/org/candlepin/subscriptions/BootApplication.java
    index 66b3a8701..c796a9581 100644
    --- a/src/main/java/org/candlepin/subscriptions/BootApplication.java
    +++ b/src/main/java/org/candlepin/subscriptions/BootApplication.java
    @@ -48,6 +48,7 @@ public class BootApplication {
         // Force liquibase-hub to off to avoid unnecessary warnings in our logs
         System.setProperty("liquibase.hub.mode", "off");

    +    System.out.println("Hello world!");
         SpringApplication app = new SpringApplication(BootApplication.class);
         app.run(args);
       }

    ```
1.  Run the live reload:
    ```
    poetry run swatchdog ee deploy
    ```
1. The script has an interactive mode where you pick the project to deploy.  Select `: <root project>`.  Then select the pod to deploy to.  You'll want to choose the `swatch-api-service` pod.  The interactive portions use [fzf](https://github.com/junegunn/fzf) so you can either type some of the name (fuzzy-matching is enabled) or just use the arrow keys.
1. The script will recompile the Java classes and rsync the changes to the running pod.

### Verification

1. The pod restarts and you'll see "Hello world!" on the console.

### CODE REVIEW

**Review by Sgitario — CHANGES_REQUESTED** [review]
The changes look promising! 
However, I tried to deploy the app using the instructions and is consistently failing with:

```
2024-02-06 08:46:13 [    INFO] [thread-74 (wait_for_ready)] [clowdapp/swatch-metrics-rhel] waiting 33sec longer
2024-02-06 08:46:13 [    INFO] [thread-73 (wait_for_ready)] [clowdapp/swatch-metrics] waiting 33sec longer
2024-02-06 08:46:13 [    INFO] [thread-79 (wait_for_ready)] [clowdapp/swatch-system-conduit] waiting 33sec longer
2024-02-06 08:46:46 [   ERROR] [thread-74 (wait_for_ready)] [clowdapp/swatch-metrics-rhel] timed out waiting for resource to be ready, details:   clowdapp/swatch-metrics-rhel not ready, status conditions:
  - DeploymentsReady: False (Deployments are not yet ready)
  - ReconciliationFailed: False
  - ReconciliationSuccessful: True
  deployment/swatch-metrics-rhel-service not ready, status conditions:
  - Available: False (Deployment does not have minimum availability.)
  - Progressing: True (ReplicaSet "swatch-metrics-rhel-service-68b6c546fb" is progressing.)

2024-02-06 08:46:46 [   ERROR] [thread-73 (wait_for_ready)] [clowdapp/swatch-metrics] timed out waiting for resource to be ready, details:   clowdapp/swatch-metrics not ready, status conditions:
  - DeploymentsReady: False (Deployments are not yet ready)
  - ReconciliationFailed: False
  - ReconciliationSuccessful: True
  deployment/swatch-metrics-service not ready, status conditions:
  - Available: False (Deployment does not have minimum availability.)
  - Progressing: True (ReplicaSet "swatch-metrics-service-75c49b5dcb" is progressing.)

2024-02-06 08:46:46 [   ERROR] [thread-79 (wait_for_ready)] [clowdapp/swatch-system-conduit] timed out waiting for resource to be ready, details:   clowdapp/swatch-system-conduit not ready, status conditions:
  - DeploymentsReady: False (Deployments are not yet ready)
  - ReconciliationFailed: False
  - ReconciliationSuccessful: True
  deployment/swatch-system-conduit-service not ready, status conditions:
  - Available: False (Deployment does not have minimum availability.)
  - Progressing: True (ReplicaSet "swatch-system-conduit-service-5bbfdbb7bb" is progressing.)

2024-02-06 08:46:46 [    INFO] [          MainThread] some resources failed to become ready: clowdapp/swatch-metrics, clowdapp/swatch-metrics-rhel, clowdapp/swatch-system-conduit
2024-02-06 08:46:47 [   ERROR] [          MainThread] hit timeout error: Could not do 'wait for all deployed resources to be ready' at /home/jcarvaja/.cache/pypoetry/virtualenvs/swatchdog--2GCRS49-py3.11/lib/python3.11/site-packages/bonfire/openshift.py:85 in time
2024-02-06 08:46:47 [    INFO] [          MainThread] releasing namespace 'ephemeral-f3y95p'
2024-02-06 08:46:48 [    INFO] [          MainThread] processing namespace reservation
2024-02-06 08:46:48 [    INFO] [          MainThread] running (pid 3874144): oc apply -f - 
2024-02-06 08:46:49 [    INFO] [         pid-3874144]  |stdout| namespacereservation.cloud.redhat.com/bonfire-reservation-9ffeb12d configured
2024-02-06 08:46:49 [    INFO] [          MainThread] releasing reservation 'bonfire-reservation-9ffeb12d' namespace 'ephemeral-f3y95p'
2024-02-06 08:46:49 [    INFO] [          MainThread] deploy failed: Could not do 'wait for all deployed resources to be ready' at /home/jcarvaja/.cache/pypoetry/virtualenvs/swatchdog--2GCRS49-py3.11/lib/python3.11/site-packages/bonfire/openshift.py:85 in time
ERROR: deploy failed: Could not do 'wait for all deployed resources to be ready' at /home/jcarvaja/.cache/pypoetry/virtualenvs/swatchdog--2GCRS49-py3.11/lib/python3.11/site-packages/bonfire/openshift.py:85 in time
```

Though I'm not sure if there is an environmental issue or is related to the changes in this PR.

Also, please add the setup information in the CONTRIBUTING.md or README.md.

**Comment by awood** [comment]
> Though I'm not sure if there is an environmental issue or is related to the changes in this PR.

Yeah, that looks to be an issue with the `bonfire deploy` not working properly.  For what it's worth, I often see `bonfire` report failure but when I go to the Openshift web console, I see that the pods are up and running.

**Comment by Sgitario** [comment]
> > Though I'm not sure if there is an environmental issue or is related to the changes in this PR.
> 
> Yeah, that looks to be an issue with the `bonfire deploy` not working properly. For what it's worth, I often see `bonfire` report failure but when I go to the Openshift web console, I see that the pods are up and running.

I did the deployment again and then interrupt the process, so bonfire does not release the namespace. I made sure that the swatch api service was up and running to move forward. 

Next, I followed the steps under the `Steps` section and when I selected the swatch-api-service pod, it failed with:

Captura de pantalla de 2024-02-06 12-06-50

Any ideas what could be happening?

**Comment by awood** [comment]
> I did the deployment again and then interrupt the process, so bonfire does not release the namespace. I made sure that the swatch api service was up and running to move forward.

Ah.  I think if you do a `reserve` ahead of time, you won't have to interrupt it.

As for your screenshot, it looks like the script could determine which container in the pod to deploy to.  Since a lot of our pods have sidecar containers, I had to add in code to handle that.  Which pod were you deploying to?  `swatch-api-service`?

**Comment by Sgitario** [comment]
> > I did the deployment again and then interrupt the process, so bonfire does not release the namespace. I made sure that the swatch api service was up and running to move forward.
> 
> Ah. I think if you do a `reserve` ahead of time, you won't have to interrupt it.
> 
> As for your screenshot, it looks like the script could determine which container in the pod to deploy to. Since a lot of our pods have sidecar containers, I had to add in code to handle that. Which pod were you deploying to? `swatch-api-service`?

Yep

**Comment by awood** [comment]
@Sgitario so my guess as to what is happening is that `swatch-api-service` perhaps didn't deploy correctly.  Maybe it was in a crash loop or something.  So the pod was still there but was "unhealthy".  And so the list of containers in the pod didn't work correctly.

Would you mind testing again with `poetry run swatchdog ee deploy --container=swatch-api-service` and pushing the root project to the `swatch-api-service-*` pod?  I updated the code to allow manually specifying the destination container.  If the command still fails, that indicates that the pod is not up and running correctly.

File: swatchdog/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  ```shell
+  poetry shell
+  ```
+* Install dependencies
+  ```shell
+  poetry install
+  ```
+* Run swatchdog
+  ```shell
+  poetry run swatchdog --version
```
Comment (Sgitario) [inline swatchdog/README.md:19]:
```suggestion
  poetry run swatchdog ee deploy
```

The --version flag is to display the swatchdog version. I think it's useful to explicitly add the "poetry run swatchdog ee deploy" command here (or add a full example section instead).

**Review by Sgitario — COMMENTED** [review]
It worked for me! I think I didn't work for me at first because I didn't select the ":" project when running the `poetry run`. 

I added a new comment in the documentation, but the rest looks good to me.

**Review by Sgitario — APPROVED** [review]

**Pull request merged**